File: Components/JsonInspect.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using Grasshopper.Kernel;
using Newtonsoft.Json.Linq;
using Enzyme.Core.UI; 

namespace Enzyme.Components
{
    public class JsonInspectComponent : GH_Component
    {
        public JsonInspectComponent()
          : base("JSON Metadata Inspector", "JSON_Inspect",
              "Extracts all unique filter keys from any MP Engine JSON.",
              "Enzyme", "Masterplan")
        {
        }

        protected override void RegisterInputParams(GH_Component.GH_InputParamManager pManager)
        {
            pManager.AddTextParameter("MP_JSON", "MP_JSON", "Any JSON stream (Masses, Slabs, Roofs, Railings, Facades).", GH_ParamAccess.item);
        }

        protected override void RegisterOutputParams(GH_Component.GH_OutputParamManager pManager)
        {
            pManager.AddTextParameter("Buildings", "Buildings", "Unique Building IDs.", GH_ParamAccess.list);
            pManager.AddTextParameter("Towers", "Towers", "Unique Tower IDs.", GH_ParamAccess.list);
            pManager.AddTextParameter("Programs", "Programs", "Unique Program names.", GH_ParamAccess.list);
            pManager.AddTextParameter("Types", "Types", "Unique element types (e.g., Roof classifications).", GH_ParamAccess.list);
            pManager.AddIntegerParameter("Levels", "Levels", "Unique floor indices, sorted numerically.", GH_ParamAccess.list);
        }

        protected override void SolveInstance(IGH_DataAccess DA)
        {
            Stopwatch stopwatch = Stopwatch.StartNew();

            string jsonIn = string.Empty;
            if (!DA.GetData(0, ref jsonIn) || string.IsNullOrWhiteSpace(jsonIn))
            {
                this.Message = this.NickName + "\nTime: 0.0 ms\n---\nAwaiting Data";
                return;
            }

            HashSet<string> bldgs = new HashSet<string>();
            HashSet<string> towers = new HashSet<string>();
            HashSet<string> progs = new HashSet<string>();
            HashSet<string> types = new HashSet<string>();
            HashSet<int> levels = new HashSet<int>();

            try
            {
                JObject data = JObject.Parse(jsonIn);

                foreach (var bldg in data)
                {
                    bldgs.Add(bldg.Key);

                    if (bldg.Value is JObject contentDict)
                    {
                        foreach (var prog in contentDict)
                        {
                            progs.Add(prog.Key);
                            if (prog.Value is JArray floors)
                            {
                                foreach (var item in floors)
                                {
                                    if (item is JObject itemDict)
                                    {
                                        if (itemDict.ContainsKey("tower_id")) towers.Add(itemDict["tower_id"].ToString());
                                        if (itemDict.ContainsKey("floor_index")) levels.Add(itemDict["floor_index"].Value<int>());
                                        if (itemDict.ContainsKey("type")) types.Add(itemDict["type"].ToString());
                                    }
                                }
                            }
                        }
                    }
                    else if (bldg.Value is JArray contentList)
                    {
                        foreach (var item in contentList)
                        {
                            if (item is JObject itemDict)
                            {
                                if (itemDict.ContainsKey("tower_id")) towers.Add(itemDict["tower_id"].ToString());
                                if (itemDict.ContainsKey("program")) progs.Add(itemDict["program"].ToString());
                                if (itemDict.ContainsKey("type")) types.Add(itemDict["type"].ToString());
                                if (itemDict.ContainsKey("floor_index")) levels.Add(itemDict["floor_index"].Value<int>());
                                
                                if (itemDict.ContainsKey("programs_above") && itemDict["programs_above"] is JArray programsAbove)
                                {
                                    foreach (var p in programsAbove)
                                    {
                                        progs.Add(p.ToString());
                                    }
                                }
                            }
                        }
                    }
                }

                List<string> outBldgs = bldgs.ToList();
                outBldgs.Sort();
                List<string> outTowers = towers.ToList();
                outTowers.Sort();
                List<string> outProgs = progs.ToList();
                outProgs.Sort();
                List<string> outTypes = types.ToList();
                outTypes.Sort();
                List<int> outLevels = levels.ToList();
                outLevels.Sort();

                stopwatch.Stop();
                double execTime = stopwatch.Elapsed.TotalMilliseconds;
                int totalTags = outBldgs.Count + outTowers.Count + outProgs.Count + outTypes.Count + outLevels.Count;

                this.Message = this.NickName + $"\nTime: {execTime:F1} ms\n---\nTags Found: {totalTags}";

                DA.SetDataList(0, outBldgs);
                DA.SetDataList(1, outTowers);
                DA.SetDataList(2, outProgs);
                DA.SetDataList(3, outTypes);
                DA.SetDataList(4, outLevels);
            }
            catch (Exception ex)
            {
                this.Message = $"JSON Parse Error:\n{ex.Message}";
            }
        }

        protected override System.Drawing.Bitmap Icon
        {
            get
            {
                return IconLoader.Load("JSON_Inspect.png");
            }
        }

        public override GH_Exposure Exposure => GH_Exposure.primary;

        public override Guid ComponentGuid => new Guid("CA44BB4A-E36D-41D1-A5BC-B0D13993D6B5");
    }
}

File: ScriptsToProcess/JSON_Inspect.py
"""
JSON METADATA INSPECTOR
================================================================================
A generic utility node that acts as an "X-Ray" for the JSON-Bus.
Reads any Masterplan Engine JSON payload and extracts all unique, filterable 
metadata tags. Used to inform the user exactly what options are available for 
downstream filtering.

INPUTS:
    MP_JSON (str) : Any JSON stream (Masses, Slabs, Roofs, Railings, Facades).

OUTPUTS:
    Buildings (str) : Unique Building IDs.
    Towers    (str) : Unique Tower IDs.
    Programs  (str) : Unique Program names.
    Types     (str) : Unique element types (e.g., Roof classifications).
    Levels    (int) : Unique floor indices, sorted numerically.
================================================================================
"""
import json
import time

# --- UI WRAPPER ---
def update_ui(msg):
    _ghenv = globals().get('ghenv')
    _comp = globals().get('component')
    if _ghenv: _ghenv.Component.Message = msg
    elif _comp: _comp.Message = msg
    else: print(msg)

_ghenv = globals().get('ghenv')
if _ghenv:
    _ghenv.Component.Name = "JSON Metadata Inspector"
    _ghenv.Component.NickName = "JSON_Inspect"
    _ghenv.Component.Description = "Extracts all unique filter keys from any MP Engine JSON."

def inspect_json():
    exec_start = time.perf_counter()
    json_in = globals().get('MP_JSON', "")
    
    # Use sets to automatically guarantee uniqueness
    bldgs = set()
    towers = set()
    progs = set()
    types = set()
    levels = set()
    
    if not json_in:
        update_ui("INSPECTOR\nTime: 0.0 ms\n---\nAwaiting Data")
        return [], [], [], [], []
        
    try:
        data = json.loads(json_in)
        
        # The top-level dictionary keys are always Building IDs
        for bldg_name, content in data.items():
            bldgs.add(str(bldg_name))
            
            # ROUTE A: Dictionary-based hierarchy (Facade JSON)
            if isinstance(content, dict):
                for prog_name, floors in content.items():
                    progs.add(str(prog_name))
                    if isinstance(floors, list):
                        for item in floors:
                            if isinstance(item, dict):
                                if "tower_id" in item: towers.add(str(item["tower_id"]))
                                if "floor_index" in item: levels.add(int(item["floor_index"]))
                                if "type" in item: types.add(str(item["type"]))
                                
            # ROUTE B: List-based hierarchy (Masses, Slabs, Roofs, Railings)
            elif isinstance(content, list):
                for item in content:
                    if isinstance(item, dict):
                        if "tower_id" in item: towers.add(str(item["tower_id"]))
                        if "program" in item: progs.add(str(item["program"]))
                        if "type" in item: types.add(str(item["type"]))
                        if "floor_index" in item: levels.add(int(item["floor_index"]))
                        
                        # Special catch for Roof JSON which uses a list of programs
                        if "programs_above" in item:
                            for p in item["programs_above"]: 
                                progs.add(str(p))
                                
        # Sort alphabetical text
        out_bldgs = sorted(list(bldgs))
        out_towers = sorted(list(towers))
        out_progs = sorted(list(progs))
        out_types = sorted(list(types))
        
        # Sort numerical levels correctly (so 2 comes before 10)
        out_levels = sorted(list(levels))
        
        exec_time = (time.perf_counter() - exec_start) * 1000
        
        # Calculate total unique tags found
        total_tags = len(out_bldgs) + len(out_towers) + len(out_progs) + len(out_types) + len(out_levels)
        update_ui(f"INSPECTOR\nTime: {exec_time:.1f} ms\n---\nTags Found: {total_tags}")
        
        return out_bldgs, out_towers, out_progs, out_types, out_levels
        
    except Exception as e:
        update_ui(f"JSON Parse Error:\n{str(e)}")
        return [], [], [], [], []

# Execute
b_list, t_list, p_list, type_list, l_list = inspect_json()

# Outputs
Buildings = b_list
Towers = t_list
Programs = p_list
Types = type_list
Levels = l_list